pkg/controller/clusterroleaggregation/clusterroleaggregation_controller_test.go

```diff
@@ -146,6 +159,11 @@ func TestSyncClusterRole(t *testing.T) {
 				indexer.Add(obj)
 			}
 			fakeClient := fakeclient.NewSimpleClientset(objs...)
+
+			// The default reactor doesn't support apply, so we need our own (trivial) reactor
```

Comment:
This looks like a problem worth fixing.

Reply:
Yeah, this is an interesting problem I hit. I chatted with a few people out-of-band about this.

Because server-side apply is implemented in the apiserver (see [fieldmanager](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager)), this looks really difficult to address (intractable). It is maybe slightly more tractable now that we have the SMD schemas loaded into the client, but still a large problem. I've opened an issue to track: https://github.com/kubernetes/kubernetes/issues/99953 with some notes about what we might be able to do (and what makes full support so difficult). Okay to look at this as a follow up to GA? I don't think we're going to be able to do much in the short term :(

cc @apelisse @liggitt

Reply:
It's difficult to recommend people use a shiny new client for their controllers if they are unable to easily unit test the code that they author.  If we cannot recommend that people use the new GA feature, is it truly GA?  Is there a pattern they can use in the interim?

Reply:
I've spent some time reviewing use cases and trying to come up with a recommendation. See [client-go Apply Fake Client: Review of use cases](https://docs.google.com/document/d/1ipi1sZd5rvPvplX_TV74gGOtFWJTf9N0qKnH2-rcI4s/edit#) for a lot more detail. There seem to be 4 main uses cases to consider:

- "Simulate responses to requests (via Reactors)" - works fine
- "Validate the requests made using the client" - I think this is where we should focus short term efforts
- "Simulate external changes to objects" - works fine
- "Using objectTracker as a pretend apiserver" - looks like an anti-pattern, found very few tests doing this

I think we can get the "Validate the requests made using the client" fixed with a relatively small change in the near term.

Reply:
I'd like to have the apply fake client be at least as good as the update fake client is today.  It certainly needs to detect when cases aren't working right as a minimum bar.